### PR TITLE
HT-126 : Create GraphQL APIs for `RoleModel`

### PR DESCRIPTION
**Pull request by Pratik1234agarwal**
Created: 
1. RoleModel in MongoDB
2. Controller to create, update, list the data in RoleModel
3. Mutation and Query in GraphQL for the model.
4. Types related to RoleModel to handle the api.

### CODE REVIEW

File: packages/api/source/controllers/RoleControllers.ts [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+
+    if (!app) {
+        throw new NotFoundError("Cannot find a Role with the specified name.");
+    }
+
+    return app;
+};
+
+const remove = async (context, name: String): Promise<{ success: Boolean }> => {
+    const role = await RoleModel.findOne({ name });
```
Comment (joelerego) [inline packages/api/source/controllers/RoleControllers.ts:123]: Needs `exec()`

File: packages/api/source/controllers/RoleControllers.ts [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+    const { error, value } = updateSchema.validate(attributes, {
+        stripUnknown: true,
+    });
+
+    if (error) {
+        throw new BadRequestError(error.message);
+    }
+
+    const app = await RoleModel.findOneAndUpdate(
+        {
```
Comment (joelerego) [inline packages/api/source/controllers/RoleControllers.ts:62]: Variable names are not changed

File: packages/api/source/controllers/RoleControllers.ts [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+            name,
+        },
+        value,
+    ).exec();
+
+    if (!app) {
+        throw new NotFoundError(
```
Comment (joelerego) [inline packages/api/source/controllers/RoleControllers.ts:69]: `app` variable name should be changed

File: packages/api/source/controllers/RoleControllers.ts [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+            "A Role Model with the specified identifier does not exist.",
+        );
+    }
+
+    return app;
+};
```
Comment (joelerego) [inline packages/api/source/controllers/RoleControllers.ts:75]: `app` variable name should be changed.

File: packages/api/source/controllers/RoleControllers.ts [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+    };
+};
+
+const listById = async (context, name: String): Promise<Role> => {
+    const filters = {
+        name,
+    };
+
+    const app = await RoleModel.findOne(filters as any).exec();
+
```
Comment (joelerego) [inline packages/api/source/controllers/RoleControllers.ts:114]: Why not just pass `{ name }`? What really is the use of filters here?

Comment (joelerego) [inline packages/api/source/controllers/RoleControllers.ts:114]: Also, this implementation is wrong. `listById` should take an array of Ids. Please do change it.

File: packages/api/source/controllers/RoleControllers.ts [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+    if (!app) {
+        throw new NotFoundError("Cannot find a Role with the specified name.");
+    }
+
+    return app;
```
Comment (joelerego) [inline packages/api/source/controllers/RoleControllers.ts:119]: Variable names.

File: packages/api/source/controllers/RoleControllers.ts [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+
+const createSchema = joi.object({
+    name: joi.string().max(512).required(),
+    priviledges: joi.array().items(joi.string().max(512).required()),
+});
+
+const updateSchema = joi.object({
+    name: joi.string(),
+    priviledges: joi.array().items(joi.string().max(512)),
+});
```
Comment (joelerego) [inline packages/api/source/controllers/RoleControllers.ts:22]: Typo isn't corrected. We spoke about this.

File: packages/api/source/controllers/RoleControllers.ts [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+    constants,
+    BadRequestError,
+    NotFoundError,
+    extractIds,
+} from "../utils";
+
+const createSchema = joi.object({
+    name: joi.string().max(512).required(),
+    priviledges: joi.array().items(joi.string().max(512).required()),
+});
```
Comment (joelerego) [inline packages/api/source/controllers/RoleControllers.ts:17]: Typo is not fixed.

File: packages/api/source/models/RoleModel.ts [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+    {
+        name: {
+            type: String,
+            required: true,
+            minlength: 1,
+            maxlength: 512,
+        },
+        priviledges: [
+            {
+                type: String,
```
Comment (joelerego) [inline packages/api/source/models/RoleModel.ts:14]: Typo here.

File: packages/api/source/graphql/privateAPI.ts [hunk 10]
```diff
@@ -535,6 +572,10 @@ const resolvers = {
 
         getQueryTemplateByName: async (parent, values, context) =>
             queryTemplates.getByName(context.request, values.name),
+        getRoleModel: async (parent, values, context) =>
+            role.list(context.request, values),
+        getRoleById: async (parent, values, context) =>
+            role.listById(context.request, values.name),
```
Comment (joelerego) [inline packages/api/source/graphql/privateAPI.ts:578]: Get rid of this query.

File: packages/api/source/types/index.ts [hunk 11]
```diff
@@ -260,10 +260,14 @@ export interface ExternalQuery {
     updatedAt: Date;
 }
 
+export interface Role {
+    user: string;
+    priviledges: string[];
+}
```
Comment (joelerego) [inline packages/api/source/types/index.ts:266]: The type has not been updated with the newer fields.

File: packages/api/source/controllers/RoleControllers.ts (hunk 8 above)
Comment (joelerego) [inline packages/api/source/controllers/RoleControllers.ts:17]: where is enabled in the `createSchema`?

File: packages/api/source/controllers/RoleControllers.ts (hunk 7 above)
Comment (joelerego) [inline packages/api/source/controllers/RoleControllers.ts:22]: The `enabled` attribute is missing in the `updateSchema`.

File: packages/api/source/controllers/RoleControllers.ts (hunk 5 above)
Comment (joelerego) [inline packages/api/source/controllers/RoleControllers.ts:114]: `app`

File: packages/api/source/graphql/privateAPI.ts [hunk 12]
```diff
@@ -199,6 +201,22 @@ const typeDefs = gql`
         records: [App!]!
     }
 
+    type RoleModel{
+        name:String!
+        priviledges:[String!]
+    }
```
Comment (joelerego) [inline packages/api/source/graphql/privateAPI.ts:207]: Typo here.

Comment (joelerego) [inline packages/api/source/graphql/privateAPI.ts:207]: `enabled` attribute.

File: packages/api/source/controllers/RoleControllers.ts (hunk 7 above)
Comment (Pratik1234agarwal) [reply packages/api/source/controllers/RoleControllers.ts:22]: Done

**Pull request closed without merging**